<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Generate embeddings. — generate_embeddings • vesalius</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Generate embeddings. — generate_embeddings"><meta name="description" content="Generate image embddings from spatial omics data."><meta property="og:description" content="Generate image embddings from spatial omics data."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">vesalius</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/vesalius.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Vesalius_Analysis/Spatial_DNA_seq.html">Vesalius: Spatial Genomics</a></li>
    <li><a class="dropdown-item" href="../articles/Vesalius_Analysis/Spatial_RNA_seq.html">Vesalius: Spatial Transcriptomics</a></li>
    <li><a class="dropdown-item" href="../articles/Vesalius_Analysis/Vesalius_MSB_analysis.html">Vesalius: MSB Manuscript Analysis</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/patrickCNMartin/Vesalius/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Generate embeddings.</h1>
      <small class="dont-index">Source: <a href="https://github.com/patrickCNMartin/Vesalius/blob/main/R/embeddings_and_tiles.R" class="external-link"><code>R/embeddings_and_tiles.R</code></a></small>
      <div class="d-none name"><code>generate_embeddings.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Generate image embddings from spatial omics data.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">generate_embeddings</span><span class="op">(</span></span>
<span>  <span class="va">vesalius_assay</span>,</span>
<span>  dim_reduction <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>  normalization <span class="op">=</span> <span class="st">"log_norm"</span>,</span>
<span>  use_counts <span class="op">=</span> <span class="st">"raw"</span>,</span>
<span>  dimensions <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  tensor_resolution <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  filter_grid <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  filter_threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  nfeatures <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_cutoff <span class="op">=</span> <span class="st">"q5"</span>,</span>
<span>  remove_lsi_1 <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-vesalius-assay">vesalius_assay<a class="anchor" aria-label="anchor" href="#arg-vesalius-assay"></a></dt>
<dd><p>vesalius_assay object.</p></dd>


<dt id="arg-dim-reduction">dim_reduction<a class="anchor" aria-label="anchor" href="#arg-dim-reduction"></a></dt>
<dd><p>string describing which dimensionality
reduction method should be used. One of the following:
"PCA", "PCA_L", "UMAP", "LSI", "LSI_UMAP".</p></dd>


<dt id="arg-normalization">normalization<a class="anchor" aria-label="anchor" href="#arg-normalization"></a></dt>
<dd><p>string describing which normalisation
method to use. One of the following "log_norm", "SCT", "TFIDF", "none".</p></dd>


<dt id="arg-use-counts">use_counts<a class="anchor" aria-label="anchor" href="#arg-use-counts"></a></dt>
<dd><p>string describing which counts should be used for the
generating emebddings. Default = "raw". See details.</p></dd>


<dt id="arg-dimensions">dimensions<a class="anchor" aria-label="anchor" href="#arg-dimensions"></a></dt>
<dd><p>numeric describing the number of Principle Components or
Latent space dimension to use. Default dimensions = 30</p></dd>


<dt id="arg-tensor-resolution">tensor_resolution<a class="anchor" aria-label="anchor" href="#arg-tensor-resolution"></a></dt>
<dd><p>numeric (range 0 - 1) describing the compression
ratio to be applied to the final image. Default = 1</p></dd>


<dt id="arg-filter-grid">filter_grid<a class="anchor" aria-label="anchor" href="#arg-filter-grid"></a></dt>
<dd><p>numeric (range 0 - 1) size of the grid used when filtering
outlier beads. Defined as a proportion of total image size. Default = 0.1</p></dd>


<dt id="arg-filter-threshold">filter_threshold<a class="anchor" aria-label="anchor" href="#arg-filter-threshold"></a></dt>
<dd><p>numeric (range 0 -1) describing the quantile
threshold at which tiles should be retained (seed details)</p></dd>


<dt id="arg-nfeatures">nfeatures<a class="anchor" aria-label="anchor" href="#arg-nfeatures"></a></dt>
<dd><p>numeric describing the number of variable features to use.</p></dd>


<dt id="arg-features">features<a class="anchor" aria-label="anchor" href="#arg-features"></a></dt>
<dd><p>custom set of features to generate embeddings</p></dd>


<dt id="arg-min-cutoff">min_cutoff<a class="anchor" aria-label="anchor" href="#arg-min-cutoff"></a></dt>
<dd><p>only used when dimensionality reduction method is
LSI or LSI_UMAP
cutoff for feature to be included in the VariableFeatures for the object.</p></dd>


<dt id="arg-remove-lsi-">remove_lsi_1<a class="anchor" aria-label="anchor" href="#arg-remove-lsi-"></a></dt>
<dd><p>logical only used when dimensionality reduction
method is LSI or LSI_UMAP
indicating if the first LSI component should be removed from further analysis
as it usually captures sequencing depth (technical variation)</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>logical output progress message or not. Default = TRUE</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>vesalius_assay</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The core principle behind vesalius is to convert spatial omics
data into an image. <code>generate_embeddings</code> allows you to convert
your omics data into a stack of gray scale images.
The stack hight will be equal to the number of dimenions you
selected excluding UMAP embeddings as only 3 dimesnions are availbale.</p>
<p>If tiles have not yet been generated (see <code><a href="generate_tiles.html">generate_tiles</a></code>),
pixel will be generated from the spatial coordinates provided.
If tiles are already present, `generate_embeddings` will skip the
tile creation step.</p>
<p>The algorithm is broadly applicable to many spatial omics assays.
Vesalius provides a 3 nornalization methods log_norm, SCTransform,
and TF-IDF.</p>
<p>The `use_counts` argument specifies which count matrix should be used
for normalization. This argument is only necessary if you use a custom
normalised count matrix. In this case, set this argument to the name
you gave your count matrix (see <code><a href="add_counts.html">add_counts</a></code>) and
`generate_embeddings` will skip the normalization and use your custom
count matrix to generate image embeddings.</p>
<p>Note that it is also possible to add custom embeddings by using the
<code><a href="add_embeddings.html">add_embeddings</a></code> function.</p>
<p>Embeddings can be created using a custom set of genes. These genes
can be provided to the `features` argument in the form of a character
vector. Note that this will not filter the count matrix hence you
will still have access to the whole matrix for downsteam analysis.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">vesalius</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># First we build a simple object</span></span></span>
<span class="r-in"><span><span class="va">ves</span> <span class="op">&lt;-</span> <span class="fu">build_vesalius_object</span><span class="op">(</span><span class="va">coordinates</span>, <span class="va">counts</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># We can do a simple run </span></span></span>
<span class="r-in"><span><span class="va">ves</span> <span class="op">&lt;-</span> <span class="fu">generate_embeddings</span><span class="op">(</span><span class="va">ves</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># maybe we want to try a different method</span></span></span>
<span class="r-in"><span><span class="co"># both will be stored in the object</span></span></span>
<span class="r-in"><span><span class="va">ves</span> <span class="op">&lt;-</span> <span class="fu">generate_embeddings</span><span class="op">(</span><span class="va">ves</span>, dim_reduction <span class="op">=</span> <span class="st">"UMAP"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Patrick C.N. Martin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

