<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Aling and integrate spatial assay from the same modality using super pixels — map_assays • vesalius</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Aling and integrate spatial assay from the same modality using super pixels — map_assays"><meta name="description" content="Aling and integrate spatial assay from the same modality using super pixels"><meta property="og:description" content="Aling and integrate spatial assay from the same modality using super pixels"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">vesalius</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/vesalius.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Vesalius_Analysis/Spatial_DNA_seq.html">Vesalius: Spatial Genomics</a></li>
    <li><a class="dropdown-item" href="../articles/Vesalius_Analysis/Spatial_RNA_seq.html">Vesalius: Spatial Transcriptomics</a></li>
    <li><a class="dropdown-item" href="../articles/Vesalius_Analysis/Vesalius_MSB_analysis.html">Vesalius: MSB Manuscript Analysis</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/patrickCNMartin/Vesalius/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Aling and integrate spatial assay from the same modality using super pixels</h1>
      <small class="dont-index">Source: <a href="https://github.com/patrickCNMartin/Vesalius/blob/main/R/map_assays.R" class="external-link"><code>R/map_assays.R</code></a></small>
      <div class="d-none name"><code>map_assays.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Aling and integrate spatial assay from the same modality using super pixels</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">map_assays</span><span class="op">(</span></span>
<span>  <span class="va">seed_assay</span>,</span>
<span>  <span class="va">query_assay</span>,</span>
<span>  signal <span class="op">=</span> <span class="st">"variable_features"</span>,</span>
<span>  use_cost <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"feature"</span>, <span class="st">"niche"</span><span class="op">)</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>  neighborhood <span class="op">=</span> <span class="st">"knn"</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  radius <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  depth <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">30</span><span class="op">)</span>,</span>
<span>  batch_size <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  epochs <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  allow_duplicates <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  use_norm <span class="op">=</span> <span class="st">"raw"</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  custom_cost <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seed_territory_labels <span class="op">=</span> <span class="st">"Territory"</span>,</span>
<span>  query_territory_labels <span class="op">=</span> <span class="st">"Territory"</span>,</span>
<span>  seed_cell_labels <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  query_cell_labels <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  jitter <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-seed-assay">seed_assay<a class="anchor" aria-label="anchor" href="#arg-seed-assay"></a></dt>
<dd><p>vesalius_assay object - data to be mapped to</p></dd>


<dt id="arg-query-assay">query_assay<a class="anchor" aria-label="anchor" href="#arg-query-assay"></a></dt>
<dd><p>vesalius_assay objecy - data to map</p></dd>


<dt id="arg-signal">signal<a class="anchor" aria-label="anchor" href="#arg-signal"></a></dt>
<dd><p>character (variable_features, all_features, embeddings, custom)
- What should  be used as cell signal to generate the cost matrix.
Seed details</p></dd>


<dt id="arg-use-cost">use_cost<a class="anchor" aria-label="anchor" href="#arg-use-cost"></a></dt>
<dd><p>character string defining how should total cost be computer
Available: feature, niche, territory, composition (See details for combinations
and custom matrices)</p></dd>


<dt id="arg-neighborhood">neighborhood<a class="anchor" aria-label="anchor" href="#arg-neighborhood"></a></dt>
<dd><p>character - how should the neighborhood be selected?
"knn", "radius", "graph"(See details)</p></dd>


<dt id="arg-k">k<a class="anchor" aria-label="anchor" href="#arg-k"></a></dt>
<dd><p>int ]2, n_points] number of neareset neighbors to be considered for
neighborhodd computation.</p></dd>


<dt id="arg-radius">radius<a class="anchor" aria-label="anchor" href="#arg-radius"></a></dt>
<dd><p>numeric ]0,1[ proportion of max distance between points
to consider for the neighborhood</p></dd>


<dt id="arg-depth">depth<a class="anchor" aria-label="anchor" href="#arg-depth"></a></dt>
<dd><p>int [1, NA] graph depth from cell to consider from neighborhood
(See details)</p></dd>


<dt id="arg-dimensions">dimensions<a class="anchor" aria-label="anchor" href="#arg-dimensions"></a></dt>
<dd><p>Int vector containing latent space dimensions to use</p></dd>


<dt id="arg-batch-size">batch_size<a class="anchor" aria-label="anchor" href="#arg-batch-size"></a></dt>
<dd><p>number of points per batch in query during assignment
problem solving</p></dd>


<dt id="arg-use-norm">use_norm<a class="anchor" aria-label="anchor" href="#arg-use-norm"></a></dt>
<dd><p>character - which count data to use</p></dd>


<dt id="arg-scale">scale<a class="anchor" aria-label="anchor" href="#arg-scale"></a></dt>
<dd><p>logical - should signal be scaled</p></dd>


<dt id="arg-threshold">threshold<a class="anchor" aria-label="anchor" href="#arg-threshold"></a></dt>
<dd><p>score threshold below which indicices should be removed.
Scores will always be between 0 and 1</p></dd>


<dt id="arg-custom-cost">custom_cost<a class="anchor" aria-label="anchor" href="#arg-custom-cost"></a></dt>
<dd><p>matrix - matrix of size n (query cells) by p (seed cells)
containing custom cost matrix. Used instead of vesalius cost matrix</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>logical - should I be a noisy boy?</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>vesalius_assay</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The goal is to assign the best matching point between a seed set and
a query set.</p>
<p>To do so, <code>map_assays</code> will first extract a
biological signal. This can be latent space embeddings per cell, or by using
gene counts (or any other modality).</p>
<p>If using gene counts, there are a few more options available to
you. First, you can select "variable_features" and vesalius will find the
intersection between the variable features in your seed_assay and your
query_assay. "all_features" will find the intersection of all genes across
assays (even if they are not highly variable). Finally, you can also select
a custom gene vector, containing only the gene set you are interested in.</p>
<p>The second step is to create a cost matrix. The creation of a cost matrix
is achieved by pair-wise sum of various cost matrices. By default,
the map_assays function will use "feature" and "niche" cost matrices.
The feature matrix computes the pearson correlation between the seed and query
using which ever signal was defined by the signal argument (variable_features)
will compute the correlation between shared variable features in seed
and query).
The niche matrix will be computed by using the pearson correlation between
niche expression profiles (based on signal). Niche are defined using the
neighborhood argument where knn represent the k nearest neighbors algorithm
(with k defining the number of nearest neighbors), depth represents the
graph depth of a local neighborhood graph, and radius defining a spatial
radius surrunding a center cell. The singal (expression or embedding) is
average across all cells in the niche.
The territory matrix will compare the average signal of vesalius
territories between seed and query.
The composition matrix will compute a frequency aware jaccard index
between cell types present in a niche. Cell types must be assigned
to seed and query vesalius objects  (See add_cells function)
Total cost matrix will be computed by computing the pairwise sum
of the complement (1 - p ) of each cost matrix.</p>
<p>This cost matrix is then parsed to a
Kuhn–Munkres algorithm that will generate point pairs that minimize
the overall cost.</p>
<p>Since the algorithm complexity is O(n3), it can be time consuming to
to run on larger data sets. As such, mapping will be approximated by
dividing seed and query into batches defined by batch size. For an
exact mapping ensure that batch_size is larger than the number of cells
in both query and seed.</p>
<p>Finaly once the matches are found, the coordinates are mapped to its
corresponding point and a new object is returned.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">vesalius</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Create Vesalius object for processing</span></span></span>
<span class="r-in"><span><span class="va">vesalius</span> <span class="op">&lt;-</span> <span class="fu"><a href="build_vesalius_assay.html">build_vesalius_assay</a></span><span class="op">(</span><span class="va">coordinates</span>, <span class="va">counts</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">jitter_ves</span> <span class="op">&lt;-</span> <span class="fu"><a href="build_vesalius_assay.html">build_vesalius_assay</a></span><span class="op">(</span><span class="va">jitter_coord</span>, <span class="va">jitter_counts</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mapped</span> <span class="op">&lt;-</span> <span class="fu">map_assays</span><span class="op">(</span><span class="va">vesalius</span>, <span class="va">jitter_ves</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Patrick C.N. Martin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

