<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Vesalius: Spatial Genomics • vesalius</title>
<script src="../../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../../deps/headroom-0.11.0/headroom.min.js"></script><script src="../../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../../deps/search-1.0.0/fuse.min.js"></script><script src="../../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../../pkgdown.js"></script><meta property="og:title" content="Vesalius: Spatial Genomics">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../../index.html">vesalius</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../../articles/vesalius.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../../articles/Vesalius_Analysis/Spatial_DNA_seq.html">Vesalius: Spatial Genomics</a></li>
    <li><a class="dropdown-item" href="../../articles/Vesalius_Analysis/Spatial_RNA_seq.html">Vesalius: Spatial Transcriptomics</a></li>
    <li><a class="dropdown-item" href="../../articles/Vesalius_Analysis/Vesalius_MSB_analysis.html">Vesalius: MSB Manuscript Analysis</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/patrickCNMartin/Vesalius/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Vesalius: Spatial Genomics</h1>
                        <h4 data-toc-skip class="author">Patrick C.N.
Martin</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/patrickCNMartin/Vesalius/blob/main/vignettes/Vesalius_Analysis/Spatial_DNA_seq.Rmd" class="external-link"><code>vignettes/Vesalius_Analysis/Spatial_DNA_seq.Rmd</code></a></small>
      <div class="d-none name"><code>Spatial_DNA_seq.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="forward">Forward<a class="anchor" aria-label="anchor" href="#forward"></a>
</h2>
<p>The following notebook highlights the analysis shown in the Vesalius
2.0 manuscript. More specifcially, we show the analysis for Spatial
Genomics Slide-SeqV2 in mouse liver.</p>
<p>The aim of this notebook is two-fold:</p>
<ol style="list-style-type: decimal">
<li>Provide a clear demonstration of how the figures in the manuscript
were produced.</li>
<li>A contextual vignette that demonstrates how one could use
Vesalius</li>
</ol>
<p>You will find through out this article subsection with the prefix
“See also”. These sections will highlight link towards documentation
that will enable you to go beyond what we did with vesalius for this
analysis.</p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<div class="section level3">
<h3 id="vesalius">Vesalius<a class="anchor" aria-label="anchor" href="#vesalius"></a>
</h3>
<p>Vesalius is a tool to decipher tissue anatomy and spatial domains
from spatial omics data. To achieve this, Vesalius converts latent space
embeddings into images upon which image processing techniques can be
applied. In contrast to many other tools, Vesalius does not create a
spatially aware latent space but utilises images that <em>represent</em>
latent space to isolate tissue domains.</p>
</div>
<div class="section level3">
<h3 id="slide-dna-seq">Slide-DNA-seq<a class="anchor" aria-label="anchor" href="#slide-dna-seq"></a>
</h3>
<p><a href="https://www.nature.com/articles/s41586-021-04217-4" class="external-link">Slide-DNA-seq</a>
is a high resolution spatial genomics data set. The authors have managed
to create a high resolution assay that captures clonal diversity by
measuring Copy Number Variation at single cell level. As it is the case
with <a href="https://www.nature.com/articles/s41587-020-0739-1" class="external-link">Slide-RNA-seq</a>,
you should keep in mind:</p>
<p><em>Spatially Indexed beads will not necessarily line up with the
cells from the tissue. As a consequence, each bead can recover DNA from
multiple cells. Conversly, in the case of large cells, transcripts might
be spread over multiple spatial indices.</em></p>
<p>This data set also provides complementary Slide-RNA-seq data. Taken
together, these data sets enable an in dpeth exploration of clonal and
transcriptonal landscapes.</p>
<p>It should be noted that the Slide-DNA-seq data comes in 3 files:</p>
<ol style="list-style-type: decimal">
<li>A count matrix</li>
<li>A coordinate file</li>
<li>A bin file that is used to link spatial indices in the count matrix
to their spatial coordinates</li>
</ol>
</div>
<div class="section level3">
<h3 id="aims">Aims<a class="anchor" aria-label="anchor" href="#aims"></a>
</h3>
<p>Vesalius aims to recover the finer details of spatial patterning by
providing uniform tissue territories.</p>
<p>Uniform territories enable:</p>
<ol style="list-style-type: decimal">
<li>In depth analysis of spatial patterning.</li>
<li>Differential gene expression analysis between territories but also
cell types present in different territories.</li>
<li>Tissue border expression.</li>
<li>Morphology dependant expression.</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="spatial-genomics-analysis">Spatial genomics analysis<a class="anchor" aria-label="anchor" href="#spatial-genomics-analysis"></a>
</h2>
<p>Before starting, make sure you have vesalius installed and that you
have your data ready and downloaded.</p>
<p>To facilitate the analysis process, we use input/output folder to
store and save data. For visualization purposes, we also load a few
extra packages. We also load the <code>future</code> package to limit
the number of cores being used.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load libraries and set seed</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patrickcnmartin.github.io/Vesalius/" class="external-link">vesalius</a></span>, lib.loc <span class="op">=</span> <span class="st">"/common/martinp4/R"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/heliosdrm/pwr" class="external-link">pwr</a></span>, lib.loc <span class="op">=</span> <span class="st">"/common/martinp4/R"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">ggpubr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1514</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Using Multi core !!! See note below !!!</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multicore</span>, workers <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Input data directory slide-DNA-seq</span></span>
<span><span class="va">spatial_coordinates_dna</span> <span class="op">&lt;-</span> <span class="st">"/common/martinp4/SSv2_DNA/mouse_liver_met_2_dna_200114_10.bead_locations.csv"</span></span>
<span><span class="va">counts_dna</span> <span class="op">&lt;-</span> <span class="st">"/common/martinp4/SSv2_DNA/mouse_liver_met_2_dna_200114_10.sparse_counts_1Mb.txt"</span></span>
<span><span class="va">bins</span> <span class="op">&lt;-</span> <span class="st">"/common/martinp4/SSv2_DNA/mm10_1Mb_bins.txt"</span></span>
<span><span class="co"># Input data directory slide-RNA-seq</span></span>
<span><span class="va">spatial_coordinates_rna</span> <span class="op">&lt;-</span> <span class="st">"/common/martinp4/SSv2/mouse_liver_met_2_rna_201002_04.bead_locations.csv"</span></span>
<span><span class="va">counts_rna</span> <span class="op">&lt;-</span> <span class="st">"/common/martinp4/SSv2/mouse_liver_met_2_rna_201002_04.sparse_expression.txt"</span></span>
<span></span>
<span><span class="co"># Output directory</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="st">"/common/martinp4/output_plots/"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"/common/martinp4/output_plots/"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">output_plots</span> <span class="op">&lt;-</span> <span class="st">"/common/martinp4/output_plots/"</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="st">"/common/martinp4/output_data/"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"/common/martinp4/output_data/"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">output_data</span> <span class="op">&lt;-</span> <span class="st">"/common/martinp4/output_data/"</span></span></code></pre></div>
<p><strong>NOTE</strong> Here, we use multicore processing from the <a href="https://cran.r-project.org/web/packages/future/index.html" class="external-link">future</a>
and <a href="https://cran.r-project.org/web/packages/future.apply/index.html" class="external-link">future.apply</a>
packages.</p>
<p>To use a single core, use <code>plan(sequential)</code></p>
<p>If you are using a windows machine, multicore is not supported.
Instead you can use <code>plan(multisession, workers = 10)</code></p>
<div class="section level3">
<h3 id="loading-slide-dna-seq">Loading Slide-DNA-seq<a class="anchor" aria-label="anchor" href="#loading-slide-dna-seq"></a>
</h3>
<p>Both of the slide-DNA and slide-RNA comme in sparse format. We will
start with the Slide-DNA-seq analysis.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spatial_coordinates_dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">spatial_coordinates_dna</span>,</span>
<span>    header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">","</span><span class="op">)</span></span>
<span><span class="va">counts_dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">counts_dna</span>,</span>
<span>    header <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">counts_dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/sparseMatrix.html" class="external-link">sparseMatrix</a></span><span class="op">(</span>i <span class="op">=</span> <span class="va">counts_dna</span><span class="op">$</span><span class="va">V2</span>,</span>
<span>  j <span class="op">=</span> <span class="va">counts_dna</span><span class="op">$</span><span class="va">V1</span>,</span>
<span>  x <span class="op">=</span> <span class="va">counts_dna</span><span class="op">$</span><span class="va">V3</span><span class="op">)</span></span>
<span><span class="va">bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">bins</span>,</span>
<span>    header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts_dna</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">spatial_coordinates_dna</span><span class="op">$</span><span class="va">barcodes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">counts_dna</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">bins</span><span class="op">$</span><span class="va">bin_ind</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">spatial_coordinates_dna</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bins</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">counts_dna</span><span class="op">)</span></span></code></pre></div>
<p>Once the data is loaded, we can create a <code>vesalius_assay</code>
object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vesalius_dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/build_vesalius_assay.html">build_vesalius_assay</a></span><span class="op">(</span>coordinates <span class="op">=</span> <span class="va">spatial_coordinates_dna</span>,</span>
<span>  counts <span class="op">=</span> <span class="va">counts_dna</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"spatial_genomics"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vesalius_dna</span></span></code></pre></div>
<div class="section level4">
<h4 id="see-also---vesalius-assays">See also - Vesalius assays<a class="anchor" aria-label="anchor" href="#see-also---vesalius-assays"></a>
</h4>
<p><code>vesalius_assay</code> objects require as minimal input the
spatial coordinates.</p>
<ul>
<li>
<strong>To add custom counts</strong> see <a href="https://patrickcnmartin.github.io/Vesalius/reference/add_counts" class="external-link">add_counts</a>.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="generate-embeddings">Generate embeddings<a class="anchor" aria-label="anchor" href="#generate-embeddings"></a>
</h3>
<p>In the context of Slide-DNA-seq and copy number variation, we expect
to find 2 copies of each “fragment” per cell in diploid organisms. Less
would signify loss of a fragment (Deletions or Mutations) and more would
signify duplication events.</p>
<p>However, since Slide-DNA-seq does not garuantee a perfect overlap
between cells and spatially resolved indices, these values might
fluctuate more significantly. This has been extesively explore in the
original <a href="https://www.nature.com/articles/s41586-021-04217-4" class="external-link">Slide-DNA-seq</a>
paper.</p>
<p>In general, this means that count value will be much lower than in
transcriptomic data. To account for this change, we recommend using
<code>TFIDF</code> and <code>LSI</code> for normalisation and generating
embeddings. These methods were developped for (Single-Cell chromatin
state)[<a href="https://www.nature.com/articles/s41592-021-01282-5" class="external-link uri">https://www.nature.com/articles/s41592-021-01282-5</a>]
analysis and are better suited for super sparse data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vesalius_dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/generate_embeddings.html">generate_embeddings</a></span><span class="op">(</span><span class="va">vesalius_dna</span>,</span>
<span>    dim_reduction <span class="op">=</span> <span class="st">"LSI"</span>,</span>
<span>    normalisation <span class="op">=</span> <span class="st">"TFIDF"</span>,</span>
<span>    tensor_resolution <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vesalius_dna</span></span></code></pre></div>
<div class="section level4">
<h4 id="see-also---generating-embeddings">See also - Generating embeddings<a class="anchor" aria-label="anchor" href="#see-also---generating-embeddings"></a>
</h4>
<p>The <code>generate_embeddings</code> function generates tiles from
spatial indices, pre-processes the count matrix (Finding variable
features and normalisation) , and generates latent space embeddings used
in the vesalius image stack.</p>
<ul>
<li>
<strong>To add custom counts</strong> see <a href="https://patrickcnmartin.github.io/Vesalius/reference/add_counts" class="external-link">add_counts</a>.</li>
<li>
<strong>To add custom embeddings</strong> see <a href="https://patrickcnmartin.github.io/Vesalius/reference/add_embeddings" class="external-link">add_embeddings</a>.</li>
<li>
<strong>To only generate tiles</strong> see<a href="https://patrickcnmartin.github.io/Vesalius/reference/generate_tiles" class="external-link">generate_tiles</a>
</li>
<li>
<strong>To run multiple embeddings</strong>, see <a href="https://patrickcnmartin.github.io/Vesalius/articles/vesalius.html" class="external-link">user
guide</a>
</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="image-prcessing">Image prcessing<a class="anchor" aria-label="anchor" href="#image-prcessing"></a>
</h3>
<p>The core concept of Vesalius is to provide spatial domains by
utilizing image processing techniques. In contrast to other methods,
Vesalius does not provide spatially aware latent space by default. It
utilises Principal Component Analsyis (as well as UMAP or SVD/LSI) to
generate a reduced dimesnionaility space.</p>
<p>Each latent space dimension will contain spatially expression
patterns and we can visualise this through the <code>image_plot</code>
function.</p>
<p><code>{ r viz_grey_dna, eval = FALSE, echo = TRUE} lsi_embedding_dna &lt;- vector("list",  30) for (i in seq(1, 30)){     lsi_embedding_dna[[i]] &lt;- image_plot(vesalius_dna,         embedding = "last",         dimensions = i) } print(ggarrange(plotlist = lsi_embedding_dna), ncol = 5)</code></p>
<p>The first thing that becomes apparent is that the majority of LSI
dimesnions do not contain any informative information. Only 1st three
dimensions would be useful to extract territories. With Vesalius, you
can specify which dimensions should be used for analysis.</p>
<p>We will run the image processing only on the first 3 dimensions.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vesalius_dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/regularise_image.html">regularise_image</a></span><span class="op">(</span><span class="va">vesalius_dna</span>,</span>
<span>    dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vesalius_dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/equalize_image.html">equalize_image</a></span><span class="op">(</span><span class="va">vesalius_dna</span>,</span>
<span>    dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>    sleft <span class="op">=</span> <span class="fl">5</span>, sright <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vesalius_dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/smooth_image.html">smooth_image</a></span><span class="op">(</span><span class="va">vesalius_dna</span>,</span>
<span>    dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>    method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"iso"</span>, <span class="st">"box"</span><span class="op">)</span>,</span>
<span>    sigma <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    box <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    iter <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>We can viszualise each dimension in the image tensor after image
processing.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lsi_embedding_dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="st">"list"</span>,  <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">lsi_embedding_dna</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/image_plot.html">image_plot</a></span><span class="op">(</span><span class="va">vesalius_dna</span>,</span>
<span>        embedding <span class="op">=</span> <span class="st">"last"</span>,</span>
<span>        dimensions <span class="op">=</span> <span class="va">i</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">lsi_embedding_dna</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p>In vesalius, each latent space dimension is handled independantly in
the form of a grey scale image. We can use this image stack to generate
isolated spatial domains and recover subtle expression patterns. The
selection of image processing parameters is dependant on the data set
used and on the biological question at hand. You might want to have more
or less territory granularity. However, the ability to visualize PCs
greatly aids is selecting the parameters that fits your needs.</p>
<div class="section level4">
<h4 id="see-also---image-processing">See also - image processing<a class="anchor" aria-label="anchor" href="#see-also---image-processing"></a>
</h4>
<p>Image prcessing is iteratively applied to the active embedding.</p>
<ul>
<li>
<strong>Understanding active embeddings in vesalius</strong> see <a href="https://patrickcnmartin.github.io/Vesalius/articles/vesalius.html" class="external-link">the
user guide</a>
</li>
<li>
<strong>Selecting different Dimensions</strong> see <a href="https://patrickcnmartin.github.io/Vesalius/articles/vesalius.html" class="external-link">the
user guide</a>
</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="image-segmentation-and-territory-isolation">Image segmentation and territory isolation<a class="anchor" aria-label="anchor" href="#image-segmentation-and-territory-isolation"></a>
</h3>
<p>The isolation of territories is achieved by a combination of image
segmentation and 2D isolation of image segments. By default, we use
<em>kmeans</em> for image segmentation. This produces a series of image
segments that can then be further isolated if they are sperated in 2D
space.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vesalius_dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/segment_image.html">segment_image</a></span><span class="op">(</span><span class="va">vesalius_dna</span>,</span>
<span>    dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"kmeans"</span>,</span>
<span>    col_resolution <span class="op">=</span> <span class="fl">3</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vesalius_dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/isolate_territories.html">isolate_territories</a></span><span class="op">(</span><span class="va">vesalius_dna</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>In this instance, we select only the 1st three dimesnions (note that
this is actually the default number of dimensions). The
<code>col_resolution</code> argument represents the number of clusters
to parse to <code>kmeans</code> however this does not necissarily define
the final number of territories. These segments will be subdivided into
small segments based on the distribution in space.</p>
<div class="section level4">
<h4 id="see-also---image-segmentation-territory-isolation">See also - Image segmentation &amp; Territory Isolation<a class="anchor" aria-label="anchor" href="#see-also---image-segmentation-territory-isolation"></a>
</h4>
<p>By default, vesalius uses kmeans clustering as it provides a simple
and effective way to segment images.</p>
<ul>
<li>
<strong>Using louvain based segmentation</strong> see the <a href="https://patrickcnmartin.github.io/Vesalius/reference/image_segmentation" class="external-link">image_segmentation
function</a>
</li>
<li>
<strong>Using leiden based segmentation</strong> see the <a href="https://patrickcnmartin.github.io/Vesalius/reference/image_segmentation" class="external-link">image_segmentation
function</a>
</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="visualization-of-territories">Visualization of territories<a class="anchor" aria-label="anchor" href="#visualization-of-territories"></a>
</h3>
<p>We can visualise the territories obtain in Slide-DNA-seq.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dna_ter</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/territory_plot.html">territory_plot</a></span><span class="op">(</span><span class="va">vesalius_dna</span><span class="op">)</span> <span class="op">+</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Clones"</span><span class="op">)</span></span>
<span><span class="va">dna_ter</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="transcriptional-diversity-within-clones">Transcriptional diversity within clones<a class="anchor" aria-label="anchor" href="#transcriptional-diversity-within-clones"></a>
</h2>
<div class="section level3">
<h3 id="slide-rna-quick-analysis">Slide-RNA quick analysis<a class="anchor" aria-label="anchor" href="#slide-rna-quick-analysis"></a>
</h3>
<p>As mentioned previosuly, the authors of Slide-DNA-seq also provide a
Slide-RNA-seq data set taken from an adjacent tissue slice. The clonal
diversity shows 2 clones (territory 2 and 3) and some healthy tissue
(territory 1).</p>
<p>Using this information, we analyse the Slide-RNA data and investigate
transcriptional changes with respect to cancer clones.</p>
<p>For this, we will process Slide-RNA-seq using the 1st 16 PCs:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spatial_coordinates_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">spatial_coordinates_rna</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">spatial_coordinates_rna</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"barcodes"</span>, <span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span></span>
<span><span class="va">counts_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">counts_rna</span>, sep <span class="op">=</span> <span class="st">","</span>, header <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">counts_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/sparseMatrix.html" class="external-link">sparseMatrix</a></span><span class="op">(</span>i <span class="op">=</span> <span class="va">count</span><span class="op">$</span><span class="va">V1</span>, j <span class="op">=</span> <span class="va">count</span><span class="op">$</span><span class="va">V2</span>, x <span class="op">=</span> <span class="va">count</span><span class="op">$</span><span class="va">V3</span><span class="op">)</span></span>
<span><span class="va">bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">input</span>, <span class="st">"SSv2/mouse_genes.txt"</span><span class="op">)</span>,</span>
<span>    header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts_rna</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">spatial_coordinates_rna</span><span class="op">$</span><span class="va">barcodes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">counts_rna</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">bins</span><span class="op">$</span><span class="va">Var1</span></span>
<span></span>
<span><span class="co">## Building vesalius assay</span></span>
<span><span class="va">vesalius_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/build_vesalius_assay.html">build_vesalius_assay</a></span><span class="op">(</span>coordinates <span class="op">=</span> <span class="va">spatial_coordinates_rna</span>,</span>
<span>    counts <span class="op">=</span> <span class="va">counts_rna</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"Spatial_Transcriptomics_cancer"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Building vesalius embeddings</span></span>
<span><span class="va">vesalius_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/generate_embeddings.html">generate_embeddings</a></span><span class="op">(</span><span class="va">vesalius_rna</span>,</span>
<span>    dim_reduction <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>    tensor_resolution <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Image processing</span></span>
<span><span class="va">vesalius_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/regularise_image.html">regularise_image</a></span><span class="op">(</span><span class="va">vesalius_rna</span>,</span>
<span>    dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">16</span><span class="op">)</span>,</span>
<span>    lambda <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    embedding <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vesalius_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/equalize_image.html">equalize_image</a></span><span class="op">(</span><span class="va">vesalius_rna</span>,</span>
<span>    dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">16</span><span class="op">)</span>,</span>
<span>    sleft <span class="op">=</span> <span class="fl">5</span>, sright <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vesalius_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/smooth_image.html">smooth_image</a></span><span class="op">(</span><span class="va">vesalius_rna</span>,</span>
<span>    dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">16</span><span class="op">)</span>,</span>
<span>    method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"iso"</span>, <span class="st">"median"</span><span class="op">)</span>,</span>
<span>    sigma <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    box <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    iter <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">## Image segmentation</span></span>
<span><span class="va">vesalius_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/segment_image.html">segment_image</a></span><span class="op">(</span><span class="va">vesalius_rna</span>,</span>
<span>    dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">16</span><span class="op">)</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"kmeans"</span>,</span>
<span>    col_resolution <span class="op">=</span> <span class="fl">12</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## Isolate territories</span></span>
<span><span class="va">vesalius_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/isolate_territories.html">isolate_territories</a></span><span class="op">(</span><span class="va">vesalius_rna</span>,</span>
<span>  trial <span class="op">=</span> <span class="st">"last"</span>,</span>
<span>  min_spatial_index <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Patrick C.N. Martin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
