<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Vesalius: Spatial Transcriptomics • vesalius</title>
<script src="../../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../../deps/headroom-0.11.0/headroom.min.js"></script><script src="../../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../../deps/search-1.0.0/fuse.min.js"></script><script src="../../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../../pkgdown.js"></script><meta property="og:title" content="Vesalius: Spatial Transcriptomics">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../../index.html">vesalius</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../../articles/vesalius.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../../articles/Vesalius_Analysis/Spatial_DNA_seq.html">Vesalius: Spatial Genomics</a></li>
    <li><a class="dropdown-item" href="../../articles/Vesalius_Analysis/Spatial_RNA_seq.html">Vesalius: Spatial Transcriptomics</a></li>
    <li><a class="dropdown-item" href="../../articles/Vesalius_Analysis/Vesalius_MSB_analysis.html">Vesalius: MSB Manuscript Analysis</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/patrickCNMartin/Vesalius/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Vesalius: Spatial Transcriptomics</h1>
                        <h4 data-toc-skip class="author">Patrick C.N.
Martin</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/patrickCNMartin/Vesalius/blob/main/vignettes/Vesalius_Analysis/Spatial_RNA_seq.Rmd" class="external-link"><code>vignettes/Vesalius_Analysis/Spatial_RNA_seq.Rmd</code></a></small>
      <div class="d-none name"><code>Spatial_RNA_seq.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="forward">Forward<a class="anchor" aria-label="anchor" href="#forward"></a>
</h2>
<p>The following notebook highlights the analysis shown in the Vesalius
2.0 manuscript. More specifcially, we show the analysis for Spatial
Transcriptomic Slide-SeqV2 in mouse brain and mouse embryo.</p>
<p>The aim of this notebook is two-fold:</p>
<ol style="list-style-type: decimal">
<li>Provide a clear demonstration of how the figures in the manuscript
were produced.</li>
<li>A contextual vignette that demonstrates how one could use
Vesalius</li>
</ol>
<p>You will find through out this article subsection with the prefix
“See also”. These sections will highlight link towards documentation
that will enable you to go beyond what we did with vesalius for this
analysis.</p>
<p>You will also find a <em>Code Block</em> section that contains the
entire analysis in a single block of code to facilitate
copy/pasting.</p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<div class="section level3">
<h3 id="vesalius">Vesalius<a class="anchor" aria-label="anchor" href="#vesalius"></a>
</h3>
<p>Vesalius is a tool to decipher tissue anatomy and spatial domains
from spatial omics data. To achieve this, Vesalius converts latent space
embeddings into images upon which image processing techniques can be
applied. In contrast to many other tools, Vesalius does not create a
spatially aware latent space but utilises images that <em>represent</em>
latent space to isolate tissue domains.</p>
</div>
<div class="section level3">
<h3 id="slide-seq-v2">Slide-seq V2<a class="anchor" aria-label="anchor" href="#slide-seq-v2"></a>
</h3>
<p><a href="https://www.nature.com/articles/s41587-020-0739-1" class="external-link">Slide-seq
version 2</a> is a high resolution spatial trancriptomics data that uses
spatially indexed beads that hybridise with mRNA species prensent in the
tissue. As such, there are a few aspect to consider when using this type
of data:</p>
<ol style="list-style-type: decimal">
<li>Spatially Indexed beads will not necessarily line up with the cells
from the tissue. As a consequence, each bead can recover mRNA transcript
from multiple cells. Conversly, in the case of large cells, transcripts
might be spread over multiple spatial indices.</li>
<li>Slide-seq provides an unbiased representation of the transcriptomic
landscape that is limited by its sensitvity to mRNA transcript
binding.</li>
</ol>
</div>
<div class="section level3">
<h3 id="aims">Aims<a class="anchor" aria-label="anchor" href="#aims"></a>
</h3>
<p>Vesalius aims to recover the finer details of spatial patterning by
providing uniform tissue territories.</p>
<p>Uniform territories enable:</p>
<ol style="list-style-type: decimal">
<li>In depth analysis of spatial patterning.</li>
<li>Differential gene expression analysis between territories but also
cell types present in different territories.</li>
<li>Tissue border expression.</li>
<li>Morphology dependant expression.</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="spatial-transcriptomic-analysis">Spatial transcriptomic analysis<a class="anchor" aria-label="anchor" href="#spatial-transcriptomic-analysis"></a>
</h2>
<p>Before starting, make sure you have vesalius installed and that you
have your data ready and downloaded.</p>
<p>To facilitate the analysis process, we use input/output folder to
store and save data. For visualization purposes, we also load a few
extra packages. We also load the <code>future</code> package to limit
the number of cores being used.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load libraries and set seed</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">ggpubr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1514</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patrickcnmartin.github.io/Vesalius/" class="external-link">vesalius</a></span>, lib.loc <span class="op">=</span> <span class="st">"/common/martinp4/R"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/heliosdrm/pwr" class="external-link">pwr</a></span>, lib.loc <span class="op">=</span> <span class="st">"/common/martinp4/R"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Using Multi core !!! See note below !!!</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multicore</span>, workers <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Input data directory Brain</span></span>
<span><span class="va">spatial_coordinates_brain</span> <span class="op">&lt;-</span> <span class="st">"/common/wonklab/SSv2/Puck_200115_08_bead_locations.csv"</span></span>
<span><span class="va">counts_brain</span> <span class="op">&lt;-</span> <span class="st">"/common/wonklab/SSv2/Puck_200115_08.digital_expression.txt.gz"</span></span>
<span></span>
<span><span class="co"># Input data directory Embryo</span></span>
<span><span class="va">spatial_coordinates_embryo</span> <span class="op">&lt;-</span> <span class="st">"/common/wonklab/SSv2/Puck_190926_03_bead_locations.csv"</span></span>
<span><span class="va">counts_embryo</span> <span class="op">&lt;-</span> <span class="st">"/common/wonklab/SSv2/Puck_190926_03.digital_expression.txt.gz"</span></span>
<span></span>
<span><span class="co"># Output directory</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="st">"/common/martinp4/output_plots/"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"/common/martinp4/output_plots/"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">output_plots</span> <span class="op">&lt;-</span> <span class="st">"/common/martinp4/output_plots/"</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="st">"/common/martinp4/output_data/"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"/common/martinp4/output_data/"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">output_data</span> <span class="op">&lt;-</span> <span class="st">"/common/martinp4/output_data/"</span></span></code></pre></div>
<p><strong>NOTE</strong> Here, we use multicore processing from the <a href="https://cran.r-project.org/web/packages/future/index.html" class="external-link">future</a>
and <a href="https://cran.r-project.org/web/packages/future.apply/index.html" class="external-link">future.apply</a>
packages.</p>
<p>To use a single core, use <code>plan(sequential)</code></p>
<p>If you are using a windows machine, multicore is not supported.
Instead you can use <code>plan(multisession, workers = 10)</code></p>
<div class="section level3">
<h3 id="loading-data-and-building-objects">Loading data and building objects<a class="anchor" aria-label="anchor" href="#loading-data-and-building-objects"></a>
</h3>
<p>The first step is to load the spatial transcriptomic data. Generally,
spatial data comes in two seperate files.</p>
<ol style="list-style-type: decimal">
<li>Spatial coordinates: This contains the x, y (even z) coordinates of
each spatially indexed barcodes.</li>
<li>Count matrix: This contains the number of gene counts at every
spatial index.</li>
</ol>
<p>Reading in spatial coordinates.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spatial_coordinates_brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">spatial_coordinates_brain</span>,</span>
<span>  header <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  skip <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">spatial_coordinates_brain</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"barcodes"</span>, <span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spatial_coordinates_embryo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">spatial_coordinates_embryo</span>,</span>
<span>  header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">spatial_coordinates_embryo</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"barcodes"</span>, <span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">spatial_coordinates_brain</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">spatial_coordinates_embryo</span><span class="op">)</span></span></code></pre></div>
<p>Note the use of <code>skip = 1</code>. This may not be required
depending on how your data is formatted.</p>
<p>Reading in count data</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts_brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">counts_brain</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">counts_embryo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">counts_embryo</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">counts_brain</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="building-a-vesalius-assay">Building a Vesalius Assay<a class="anchor" aria-label="anchor" href="#building-a-vesalius-assay"></a>
</h3>
<p>The first step in the Vesalius Pipeline is to build a <a href="">vesalius_assay object</a>. This object is used throughout the
vesalius workflow and stores data after every round of computation.</p>
<p>To build a <code>vesalius_assay</code> :</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vesalius_brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/build_vesalius_assay.html">build_vesalius_assay</a></span><span class="op">(</span>coordinates <span class="op">=</span> <span class="va">spatial_coordinates_brain</span>,</span>
<span>  counts <span class="op">=</span> <span class="va">counts_brain</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"Spatial_transcriptomics_brain"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">vesalius_embryo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/build_vesalius_assay.html">build_vesalius_assay</a></span><span class="op">(</span>coordinates <span class="op">=</span> <span class="va">spatial_coordinates_embryo</span>,</span>
<span>  counts <span class="op">=</span> <span class="va">counts_embryo</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"Spatial_transcriptomics_embryo"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">vesalius_brain</span></span>
<span><span class="va">vesalius_embryo</span></span></code></pre></div>
<div class="section level4">
<h4 id="see-also---vesalius-assays">See also - Vesalius assays<a class="anchor" aria-label="anchor" href="#see-also---vesalius-assays"></a>
</h4>
<p><code>vesalius_assay</code> objects require as minimal input the
spatial coordinates.</p>
<ul>
<li>
<strong>To add custom counts</strong> see <a href="https://patrickcnmartin.github.io/Vesalius/reference/add_counts" class="external-link">add_counts</a>.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="generating-embeddings">Generating embeddings<a class="anchor" aria-label="anchor" href="#generating-embeddings"></a>
</h3>
<p>The next step is to generate pixels tiles from spatial coordinates
and latent space embedding that will be used to populate each tile with
a grey scale color value.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vesalius_brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/generate_embeddings.html">generate_embeddings</a></span><span class="op">(</span><span class="va">vesalius_brain</span>,</span>
<span>    dim_reduction <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>    dimensions <span class="op">=</span> <span class="fl">50</span>,</span>
<span>    tensor_resolution <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vesalius_embryo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/generate_embeddings.html">generate_embeddings</a></span><span class="op">(</span><span class="va">vesalius_embryo</span>,</span>
<span>    dim_reduction <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>    dimensions <span class="op">=</span> <span class="fl">50</span>,</span>
<span>    tensor_resolution <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vesalius_brain</span></span>
<span><span class="va">vesalius_embryo</span></span></code></pre></div>
<p>Here, we will use Principal Component Analysis (PCA) to reduce data
dimensionality and produce our latent space. We will also reduce the
dimension of the image tensor. Effectively, we are reducing the size of
the image by combining neighboring spatial indices if they fall within
the same local area. Reducing the image resolution ensures lower run
times.</p>
<p>In addition to merging spatial indices, we adjust the count matrix by
taking the average count values for all spatial indices that have been
merged together. Note that the spatial index names become a combination
of the original spatial indices.</p>
<p>For example:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"_et_"</span>,</span>
<span>  x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="../../reference/get_tiles.html">get_tiles</a></span><span class="op">(</span><span class="va">vesalius_brain</span><span class="op">)</span><span class="op">$</span><span class="va">barcodes</span><span class="op">)</span>,</span>
<span>  value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="see-also---generating-embeddings">See also - Generating embeddings<a class="anchor" aria-label="anchor" href="#see-also---generating-embeddings"></a>
</h4>
<p>The <code>generate_embeddings</code> function generates tiles from
spatial indices, pre-processes the count matrix (Finding variable
features and normalisation) , and generates latent space embeddings used
in the vesalius image stack.</p>
<ul>
<li>
<strong>To add custom counts</strong> see <a href="https://patrickcnmartin.github.io/Vesalius/reference/add_counts" class="external-link">add_counts</a>.
For more information, refer to (modifying objects vignette)[].</li>
<li>
<strong>To add custom embeddings</strong> see <a href="https://patrickcnmartin.github.io/Vesalius/reference/add_embeddings" class="external-link">add_embeddings</a>.
For more information, refer to (modifying objects vignette)[].</li>
<li>
<strong>To only generate tiles</strong> see<a href="https://patrickcnmartin.github.io/Vesalius/reference/generate_tiles" class="external-link">generate_tiles</a>
</li>
<li>
<strong>To run multiple embeddings</strong>, see <a href="https://patrickcnmartin.github.io/Vesalius/articles/vesalius.html" class="external-link">user
guide</a>
</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="image-processing">Image processing<a class="anchor" aria-label="anchor" href="#image-processing"></a>
</h3>
<p>The core concept of Vesalius is to provide spatial domains by
utilizing image processing techniques. In contrast to other methods,
Vesalius does not provide spatially aware latent space by default. It
utilises Principal Component Analsyis (as well as UMAP or SVD/LSI) to
generate a reduced dimesnionaility space.</p>
<p>We find that Principal Components (PC) contain subtle spatial
information already embedded within them. We can viszualise this by
looking at the grey scale embeddings for each PC.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_embedding_brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="st">"list"</span>,  <span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">pca_embedding_embryo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="st">"list"</span>,  <span class="fl">30</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">30</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">pca_embedding_brain</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/image_plot.html">image_plot</a></span><span class="op">(</span><span class="va">vesalius_brain</span>,</span>
<span>        embedding <span class="op">=</span> <span class="st">"last"</span>,</span>
<span>        dimensions <span class="op">=</span> <span class="va">i</span><span class="op">)</span></span>
<span>    <span class="va">pca_embedding_embryo</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/image_plot.html">image_plot</a></span><span class="op">(</span><span class="va">vesalius_embryo</span>,</span>
<span>        embedding <span class="op">=</span> <span class="st">"last"</span>,</span>
<span>        dimensions <span class="op">=</span> <span class="va">i</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>For the mouse brain:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">pca_embedding_brain</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>For the mouse embryo:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">pca_embedding_embryo</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>We can see that many of these PCs contain spatially relevant
information even though the latent space was not produced in an
explitely spatially aware manner. By using image processing, we can
further highlight these regions and better utilise them during spatial
domain isolation.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vesalius_brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/regularise_image.html">regularise_image</a></span><span class="op">(</span><span class="va">vesalius_brain</span>,</span>
<span>    dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vesalius_brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/equalize_image.html">equalize_image</a></span><span class="op">(</span><span class="va">vesalius_brain</span>,</span>
<span>    dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>    sleft <span class="op">=</span> <span class="fl">5</span>, sright <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vesalius_brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/smooth_image.html">smooth_image</a></span><span class="op">(</span><span class="va">vesalius_brain</span>,</span>
<span>    dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>    method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"iso"</span>, <span class="st">"box"</span><span class="op">)</span>,</span>
<span>    sigma <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    box <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    iter <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">vesalius_embryo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/regularise_image.html">regularise_image</a></span><span class="op">(</span><span class="va">vesalius_embryo</span>,</span>
<span>   embedding <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>   dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>   lambda <span class="op">=</span> <span class="fl">10</span>,</span>
<span>   verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vesalius_embryo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/equalize_image.html">equalize_image</a></span><span class="op">(</span><span class="va">vesalius_embryo</span>,</span>
<span>    dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>    sleft <span class="op">=</span> <span class="fl">1</span>, sright <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vesalius_embryo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/smooth_image.html">smooth_image</a></span><span class="op">(</span><span class="va">vesalius_embryo</span>,</span>
<span>    dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>    method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"iso"</span>, <span class="st">"box"</span><span class="op">)</span>,</span>
<span>    sigma <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    box <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    across_levels <span class="op">=</span> <span class="st">"min"</span>,</span>
<span>    iter <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>We can viszualise each PC in the image tensor after image
processing.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_embedding_brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="st">"list"</span>,  <span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">pca_embedding_embryo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="st">"list"</span>,  <span class="fl">30</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">30</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">pca_embedding_brain</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/image_plot.html">image_plot</a></span><span class="op">(</span><span class="va">vesalius_brain</span>,</span>
<span>        embedding <span class="op">=</span> <span class="st">"last"</span>,</span>
<span>        dimensions <span class="op">=</span> <span class="va">i</span><span class="op">)</span></span>
<span>    <span class="va">pca_embedding_embryo</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/image_plot.html">image_plot</a></span><span class="op">(</span><span class="va">vesalius_embryo</span>,</span>
<span>        embedding <span class="op">=</span> <span class="st">"last"</span>,</span>
<span>        dimensions <span class="op">=</span> <span class="va">i</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>For example, in the mous brain, we can cleary see that PC17 contains
information related to the CA2 field - a field that is often lost during
standard analysis.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">pca_embedding_brain</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>For the mouse embryo:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">pca_embedding_embryo</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>In vesalius, each PC is handled independantly in the form of a grey
scale image. We can use this image stack to generate isolated spatial
domains and recover subtle expression patterns. The selection of image
processing parameters is dependant on the data set used and on the
biological question at hand. You might want to have more or less
territory granularity. However, the ability to visualize PCs greatly
aids is selecting the parameters that fits your needs.</p>
<div class="section level4">
<h4 id="see-also---image-processing">See also - image processing<a class="anchor" aria-label="anchor" href="#see-also---image-processing"></a>
</h4>
<p>Image prcessing is iteratively applied to the active embedding.</p>
<ul>
<li>
<strong>Understanding active embeddings in vesalius</strong> see <a href="https://patrickcnmartin.github.io/Vesalius/articles/vesalius.html" class="external-link">the
user guide</a>
</li>
<li>
<strong>Selecting different PCs</strong> see <a href="https://patrickcnmartin.github.io/Vesalius/articles/vesalius.html" class="external-link">the
user guide</a>
</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="image-segmentation-territory-isolation">Image segmentation &amp; Territory Isolation<a class="anchor" aria-label="anchor" href="#image-segmentation-territory-isolation"></a>
</h3>
<p>The isolation of territories is achieved by a combination of image
segmentation and 2D isolation of image segments. By default, we use
<em>kmeans</em> for image segmentation. This produces a series of image
segments that can then be further isolated if they are sperated in 2D
space.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vesalius_brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/segment_image.html">segment_image</a></span><span class="op">(</span><span class="va">vesalius_brain</span>,</span>
<span>    dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">30</span><span class="op">)</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"kmeans"</span>,</span>
<span>    col_resolution <span class="op">=</span> <span class="fl">25</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vesalius_brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/isolate_territories.html">isolate_territories</a></span><span class="op">(</span><span class="va">vesalius_brain</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">vesalius_embryo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/segment_image.html">segment_image</a></span><span class="op">(</span><span class="va">vesalius_embryo</span>,</span>
<span>    dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">30</span><span class="op">)</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"kmeans"</span>,</span>
<span>    col_resolution <span class="op">=</span> <span class="fl">32</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vesalius_embryo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/isolate_territories.html">isolate_territories</a></span><span class="op">(</span><span class="va">vesalius_embryo</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>In this instance, we elected to use all images in the image stack
(i.e <code>dimensions = seq(1,30)</code>). In the context of vesalius,
the <code>col_resolution</code> argument represents the number of
clusters to parse to <code>kmeans</code> however this does not define
the final number of territories. These segments will be subdivided into
small segments based on the distribution in space.</p>
<div class="section level4">
<h4 id="see-also---image-segmentation-territory-isolation">See also - Image segmentation &amp; Territory Isolation<a class="anchor" aria-label="anchor" href="#see-also---image-segmentation-territory-isolation"></a>
</h4>
<p>By default, vesalius uses kmeans clustering as it provides a simple
and effective way to segment images.</p>
<ul>
<li>
<strong>Using louvain based segmentation</strong> see the <a href="https://patrickcnmartin.github.io/Vesalius/reference/image_segmentation" class="external-link">image_segmentation
function</a>
</li>
<li>
<strong>Using leiden based segmentation</strong> see the <a href="https://patrickcnmartin.github.io/Vesalius/reference/image_segmentation" class="external-link">image_segmentation
function</a>
</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="visualization-of-territories">Visualization of territories<a class="anchor" aria-label="anchor" href="#visualization-of-territories"></a>
</h3>
<p>As mentionned above we distinguish image segements and territories.
We can vizualise these differences with the vesalius plotting
functions.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">brain_segments</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/territory_plot.html">territory_plot</a></span><span class="op">(</span><span class="va">vesalius_brain</span>,</span>
<span>  trial <span class="op">=</span> <span class="st">"Segment"</span>,</span>
<span>  cex_pt <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Brain Segments"</span><span class="op">)</span></span>
<span><span class="va">brain_territories</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/territory_plot.html">territory_plot</a></span><span class="op">(</span><span class="va">vesalius_brain</span>,</span>
<span>  trial <span class="op">=</span> <span class="st">"Territory"</span>,</span>
<span>  cex_pt <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Brain Territories"</span><span class="op">)</span></span>
<span><span class="va">embryo_segments</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/territory_plot.html">territory_plot</a></span><span class="op">(</span><span class="va">vesalius_embryo</span>,</span>
<span>  trial <span class="op">=</span> <span class="st">"Segment"</span>,</span>
<span>  cex_pt <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Embryo Segments"</span><span class="op">)</span></span>
<span><span class="va">embryo_territories</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/territory_plot.html">territory_plot</a></span><span class="op">(</span><span class="va">vesalius_embryo</span>,</span>
<span>  trial <span class="op">=</span> <span class="st">"Territory"</span>,</span>
<span>  cex_pt <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Embryo Territories"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">all</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">brain_segments</span> <span class="op">+</span> <span class="va">brain_territories</span><span class="op">)</span> <span class="op">/</span></span>
<span>  <span class="op">(</span><span class="va">embryo_segments</span> <span class="op">+</span> <span class="va">embryo_territories</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">all</span><span class="op">)</span></span></code></pre></div>
<p>We can see how image segments and territories differ when compareing
their plots. While segments are defined by the value parsed to
<code>col_resolution</code> during image segmentation, the final number
of territories will depend on how the images were processed, the value
parsed to <code>col_resolution</code> and the <em>capture radius</em>
defined in <a href="https://patrickcnmartin.github.io/Vesalius/reference/isolate_territories" class="external-link">isolate_territories</a>.
In short, this parameters defines the maximum distance between color
segments for them to be pooled into the same territory.</p>
</div>
<div class="section level3">
<h3 id="differential-gene-expression">Differential Gene Expression<a class="anchor" aria-label="anchor" href="#differential-gene-expression"></a>
</h3>
<p>The advantage of uniform territories is that it enables the
comparison of spatial territories rather than comparing cell
clusters.</p>
<p>We can start by getting all genes that are differentially expressed
in each territory and export the results.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vesalius_brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/identify_markers.html">identify_markers</a></span><span class="op">(</span><span class="va">vesalius_brain</span>,</span>
<span>  trial <span class="op">=</span> <span class="st">"Territory"</span>, <span class="co"># Using PCA territories here</span></span>
<span>  method <span class="op">=</span> <span class="st">"DESeq2"</span>,</span>
<span>  seed <span class="op">=</span>  <span class="cn">NULL</span>,</span>
<span>  query <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">vesalius_embryo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/identify_markers.html">identify_markers</a></span><span class="op">(</span><span class="va">vesalius_embryo</span>,</span>
<span>  trial <span class="op">=</span> <span class="st">"Territory"</span>, <span class="co"># Using PCA territories here</span></span>
<span>  method <span class="op">=</span> <span class="st">"DESeq2"</span>,</span>
<span>  seed <span class="op">=</span>  <span class="cn">NULL</span>,</span>
<span>  query <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">deg_brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/get_markers.html">get_markers</a></span><span class="op">(</span><span class="va">vesalius_brain</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">deg_brain</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">output_data</span>, <span class="st">"DEG_SSv2_brain.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">deg_embryo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/get_markers.html">get_markers</a></span><span class="op">(</span><span class="va">vesalius_embryo</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">deg_embryo</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">output_data</span>, <span class="st">"DEG_SSv2_embryo.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">deg_brain</span><span class="op">)</span></span></code></pre></div>
<p>Using these tables, we can ivestigate differntial gene expression and
find which genes are more highly expressed in a given territory comapred
to everything else. If you want to only look at one or a subset of
territories, you can specifiy your territories in the <code>seed</code>
argument.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vesalius_brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/identify_markers.html">identify_markers</a></span><span class="op">(</span><span class="va">vesalius_brain</span>,</span>
<span>  trial <span class="op">=</span> <span class="st">"Territory"</span>, <span class="co"># Using PCA territories here</span></span>
<span>  method <span class="op">=</span> <span class="st">"DESeq2"</span>,</span>
<span>  seed <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  query <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<p>We can also use this option to compare territories between each
other. If there are many territories it can be easier to visualize each
territory seperately.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">brain_split</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/territory_plot.html">territory_plot</a></span><span class="op">(</span><span class="va">vesalius_brain</span>,</span>
<span>  trial <span class="op">=</span> <span class="st">"Territory"</span>,</span>
<span>  split <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  randomise <span class="op">=</span> <span class="cn">FALSE</span> <span class="co"># we don't reallyt need to randomise colors here</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">brain_split</span></span></code></pre></div>
<p>In this case, we can compare territory 18 and territory 3 (medial
habenula compartments).</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vesalius_brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/identify_markers.html">identify_markers</a></span><span class="op">(</span><span class="va">vesalius_brain</span>,</span>
<span>    trial <span class="op">=</span> <span class="st">"Territory"</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"DESeq2"</span>,</span>
<span>    seed <span class="op">=</span> <span class="fl">18</span>,</span>
<span>    query <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">med_comp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/get_markers.html">get_markers</a></span><span class="op">(</span><span class="va">vesalius_brain</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">med_comp</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="see-also---using-different-deg-methods">See also - using different DEG methods<a class="anchor" aria-label="anchor" href="#see-also---using-different-deg-methods"></a>
</h4>
<p>The <code>indentify_markers</code> function offers different
comparison options.</p>
<ul>
<li>
<strong>Using different DEG methods</strong> - see (identify_markers
function)[<a href="https://patrickcnmartin.github.io/Vesalius/reference/identify_markers" class="external-link uri">https://patrickcnmartin.github.io/Vesalius/reference/identify_markers</a>]</li>
<li>
<strong>Extracting markers from object</strong> - see (get_markers
function)[<a href="https://patrickcnmartin.github.io/Vesalius/reference/get_markers" class="external-link uri">https://patrickcnmartin.github.io/Vesalius/reference/get_markers</a>]</li>
<li>**Comparing cells between territories - see (User guide)[<a href="https://patrickcnmartin.github.io/Vesalius/articles/vesalius.html" class="external-link uri">https://patrickcnmartin.github.io/Vesalius/articles/vesalius.html</a>]</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="visualization-of-gene-expresssion">Visualization of Gene expresssion<a class="anchor" aria-label="anchor" href="#visualization-of-gene-expresssion"></a>
</h3>
<p>The final step is to visualise gene expression patterns.</p>
<p>For <em>Pcp4</em> in the mouse hippocampus:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcp4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/view_gene_expression.html">view_gene_expression</a></span><span class="op">(</span><span class="va">vesalius_brain</span>,</span>
<span>  genes <span class="op">=</span> <span class="st">"Pcp4"</span>,</span>
<span>  trial <span class="op">=</span> <span class="st">"Territory"</span>,</span>
<span>  cex <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> </span>
<span><span class="va">pcp4_layer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/view_gene_expression.html">view_gene_expression</a></span><span class="op">(</span><span class="va">vesalius_brain</span>,</span>
<span>  genes <span class="op">=</span> <span class="st">"Pcp4"</span>,</span>
<span>  trial <span class="op">=</span> <span class="st">"Territory"</span>,</span>
<span>  as_layer <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cex <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pcp4</span> <span class="op">+</span> <span class="va">pcp4_layer</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>tag_levels <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span></span></code></pre></div>
<p>For <em>Tac1</em> in the mouse hippocampus:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Tac1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/view_gene_expression.html">view_gene_expression</a></span><span class="op">(</span><span class="va">vesalius_brain</span>,</span>
<span>  genes <span class="op">=</span> <span class="st">"Tac1"</span>,</span>
<span>  trial <span class="op">=</span> <span class="st">"Territory"</span>,</span>
<span>  cex <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> </span>
<span><span class="va">Tac1_layer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/view_gene_expression.html">view_gene_expression</a></span><span class="op">(</span><span class="va">vesalius_brain</span>,</span>
<span>  genes <span class="op">=</span> <span class="st">"Tac1"</span>,</span>
<span>  trial <span class="op">=</span> <span class="st">"Territory"</span>,</span>
<span>  as_layer <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cex <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="va">Tac1</span> <span class="op">+</span> <span class="va">Tac1_layer</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>tag_levels <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span></span></code></pre></div>
<p>For <em>Pmel</em> in the mouse embryo:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Pmel</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/view_gene_expression.html">view_gene_expression</a></span><span class="op">(</span><span class="va">vesalius_embryo</span>,</span>
<span>  genes <span class="op">=</span> <span class="st">"Pmel"</span>,</span>
<span>  trial <span class="op">=</span> <span class="st">"Territory"</span>,</span>
<span>  cex <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> </span>
<span><span class="va">Pmel_layer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/view_gene_expression.html">view_gene_expression</a></span><span class="op">(</span><span class="va">vesalius_embryo</span>,</span>
<span>  genes <span class="op">=</span> <span class="st">"Pmel"</span>,</span>
<span>  trial <span class="op">=</span> <span class="st">"Territory"</span>,</span>
<span>  as_layer <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cex <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="va">Pmel</span> <span class="op">+</span> <span class="va">Pmel_layer</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>tag_levels <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span></span></code></pre></div>
<p>We can visualise gene expression as an overall pattern (“A” plots) or
an averge expression in each territory (“B” plots). It is important to
use “B” plots with caution. While these plots can help to clarify the
expression distribution across territories, they represent a discretised
view of expression patterns. As such many transcriptional dynamics such
as gradients or overall low expression can altered.</p>
<div class="section level4">
<h4 id="see-also---visualization-of-multiple-genes">See also - Visualization of multiple genes<a class="anchor" aria-label="anchor" href="#see-also---visualization-of-multiple-genes"></a>
</h4>
<p>You can also parse a vector of genes to the <code>genes</code>
argument. This will return a <code>ggarrange</code> object that can
directly be plotted.</p>
<ul>
<li>
<strong>Visualization of multiple genes</strong> - see <a href="https://patrickcnmartin.github.io/Vesalius/reference/view_gene_expression" class="external-link">view_gene_expression
function</a>
</li>
<li>
<strong>Understaing ggarrange</strong> - see <a href="https://cran.r-project.org/web/packages/ggpubr/ggpubr.pdf" class="external-link">ggpubr
page</a>
</li>
</ul>
</div>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>Here, we present a Vesalius workflow using 2 high resolution spatial
transcriptomic data sets. This pipeline is applicabled to any spatial
transcriptomic data set. However, <strong>image processing parameteres
should be set according to your needs and the data set!</strong></p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.1 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] digest_0.6.37     desc_1.4.3        R6_2.5.1          fastmap_1.2.0    </span></span>
<span><span class="co">##  [5] xfun_0.50         cachem_1.1.0      knitr_1.49        htmltools_0.5.8.1</span></span>
<span><span class="co">##  [9] rmarkdown_2.29    lifecycle_1.0.4   cli_3.6.3         sass_0.4.9       </span></span>
<span><span class="co">## [13] pkgdown_2.1.1     textshaping_0.4.1 jquerylib_0.1.4   systemfonts_1.2.0</span></span>
<span><span class="co">## [17] compiler_4.4.2    tools_4.4.2       ragg_1.3.3        bslib_0.8.0      </span></span>
<span><span class="co">## [21] evaluate_1.0.3    yaml_2.3.10       jsonlite_1.8.9    rlang_1.1.4      </span></span>
<span><span class="co">## [25] fs_1.6.5          htmlwidgets_1.6.4</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Patrick C.N. Martin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
